nvaccess/nvda
# Fix eSpeak language switching by checking for fallback dialects

### Link to issue number:
Closes #13727 

### Summary of the issue:
eSpeak fails to perform automatic language switching for a common language codes "en", "fr",
Reported to eSpeak project: https://github.com/espeak-ng/espeak-ng/issues/1200
A PR has been opened against eSpeak to fix this: https://github.com/espeak-ng/espeak-ng/pull/1211

### Description of how this pull request fixes the issue:

Creates a dictionary of aliases for language codes without locale,  which map to appropriate supported languages with locale.
This may need to be added to over time as appropriate.
Integration tests have been added to ensure this list is comprehensive when we update eSpeak.

When eSpeak is unaware of a language code, the new behaviour is to:
* check for a default, if not then
* check for the language stripped of the dialect locale, if not then
* check for any matching language (regardless of dialect, e.g 'en-au' to 'en-us'), if not then
* ignore the language change

### Testing strategy:

Manual testing with STR in #13727

1. Set NVDA language and eSpeak language to French
1. Open the following page in Chrome: `data:text/html,<p>Hello</p><p lang="en">world</p>`
1. Confirm "world" is spoken in an US English accent
1. Open the following page in Chrome: `data:text/html,<p>Hello</p><p lang="de-foobar">world</p>`
1. Confirm "world" is spoken in a German accent
1. Open the following page in Chrome: `data:text/html,<p>Hello</p><p lang="foobar">world</p>`
1. Confirm "world" is spoken in the French accent

Unit testing has been added to test scenarios for determining a supported language to switch to.

Integration testing has also been added to confirm the default mappings for eSpeak continue to be supported and comprehensive.

### Known issues with pull request:
None

### Change log entries:
Bug fixes
`- Fix eSpeak automatic language switching to English and French by falling back to a similar dialect if possible. (#13727)`

### Code Review Checklist:



  - description is up to date
  - change log entries
  - Unit tests
  - System (end to end) tests
  - Manual testing
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
